types/express-serve-static-core/index.d.ts

```diff
@@ -96,16 +98,23 @@ export type RequestHandlerParams<
     | ErrorRequestHandler<P, ResBody, ReqBody, ReqQuery, Locals>
     | Array<RequestHandler<P> | ErrorRequestHandler<P>>;
 
-export type RouteParameterNames<Route extends string> =
-    string extends Route
-        ? string
-        : Route extends `${string}:${infer Param}/${infer Rest}`
-        ? (Param | RouteParameterNames<Rest>)
-        : (Route extends `${string}:${infer LastParam}` ? LastParam : never);
-
-export type RouteParameters<T extends string> = {
-    [key in RouteParameterNames<T>]: string
-};
+type GetRouteParameter<RouteAfterColon extends string> = RouteAfterColon extends `${infer Char}${infer Rest}`
```

Comment:
Another important part of `path-to-regexp` is that [parameter names](https://www.npmjs.com/package/path-to-regexp#named-parameters) are limited to `[A-Za-z0-9_]`. That's a bit harder to restrict without hitting the recursion depth limiter, but an example of this might be:

```ts
type ParameterName<S extends string, Prev extends string = ""> =
    S extends `${infer Char}${infer Rest}` ?
        "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890_" extends `${string}${Char}${string}` ? ParameterName<Rest, `${Prev}${Char}`> :
        [Prev, S] :
    [Prev, S];
```

[TypeScript Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBACghgJzgWwsCCByKIB4DKUEAHugHYAmAzlFcAgJZkDmANLAhAG5GkSU06jFlAC8UAEQSAfGICwAKCjKohEuWpQABgBIA3kwBmGKAGEAFogC++oyYBKEOla1QA-IpVfJcAEYBjCghDZnMGACsAawAbZDIAezAARwQ6AFcuAHdiEAAvAEEAIVMAEQBRADEAcQAJAEkAKQBpABkAWUwAeRgARXt8ABUAVQA1AHUADQBNAC0ARgAmAGYAFgBWADYAdgAOAE4ABgB9CV4NGl09ISZmGz0La31rlhd3WEQcdCwcXEc6dkuME4XDuDwQLlkAC5PN5lABtIHcdj4AC6UGhSlhUARwORKIA3IpFHoYcpQJAoEcxNjSd4jkcBk5gLh6WViJB-MABuA8PAkKgvthULgJH5-AB6IKGGTsOGigISdgSSXBCQo6TSWkExRWIkKcXiqAAYk4zAY8TIUEZdCg5gg0UgCCg5KcigNUGQTAYzp5zqZNyghn5EEy8QQkUDCHiyFtwGAYCokINZuA5jSvgAdP5o+KnIZiLmqPnxb5ovFfOLkHA6BhxWA4P5InBmE5xS6ALTIeJBaLiqgICX09B0Kj08VMILEDPAKhGlo7NZLRQMZBgMPAKB6KB1KiYbgYdg7-JkEBQKyR6OSAACeYLHa7PYkhIULspDKZuAGZ34mi39aoibOggaTQFYsjiAMz6vqy7IQJy3KQLg+ScmkcDROwbIcugFDgbScI7nuXAYLgmFwdh0houoP40HC9AgWibjYgR+4IEhKFoRR34CNidEQAxm5QP+gG8We6ICUJkKBmhVAQBhsGchAFCSaRCkUOwyHAKh0SSRpWmiRiXj4buLFsZpHGUXw3G0cBfFvH+1aAYY0myVAKnYcp8nYep7HaVAulofpeFHieJGeYpnFUVZvH8UZx4gKZWkRZZmjWfRdmCQ5kkieekn2QBklOdEMlyVhikeaVal+T5Ok+YFmLwsF8X+dESXnDxNn8XljnOSVZFla5YWVc1NVmdEdWGW5ineaNFltXCzW9apnUZflQEgeNKhdQVPUDRV5V9UN1VVaNZ7Pu6Ro-qa5qWtaG52g6JgulQQA)

This picks a sequence of "word chars" (per `path-to-regexp`) from the start of `S` up until the first non-"word char", and returns that sequence as the first element and the remaining path as the second element:

```ts
ParameterName<"abc/def"> // ["abc", "/def"]
```

You could even modify it to pick out the optionality:

```ts
type ParameterName<S extends string, Prev extends string = ""> =
    S extends `${infer Char}${infer Rest}` ?
        "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890_" extends `${string}${Char}${string}` ? ParameterName<Rest, `${Prev}${Char}`> :
        Char extends "?" ? [Prev, true, S] :
        [Prev, false, S] :
    [Prev, false, S];
```

This suggestion comes with a caveat: Parsing strings a character at a time with a recursive type alias very quickly hits the recursion limiter. You're welcome to ignore this suggestion if it seems infeasible.

Reply:
this is really cool, but yeah i think the recursion limit is too low here.

btw i really like those testing types you made. is there a package i can import with those? i'd like to use them for my own utility types library

Reply:
I will probably just publish them under `@esfx/type-model` at some point.